File: ext/examples/unstable_chart.rs
//! Computes the suspension map between different unstable Ext groups.
//!
//! Given an unstable Steenrod module $M$, compute the unstable Ext groups of $\Sigma^k M$ for all
//! $k$ up till the stable range. Each result is printed in the form
//! ```
//! n s min_degree: num_gens - matrix
//! ```
//! The entries are to be interpreted as follows:
//!  - `n` is the stem, which is defined to be `t - s - min_degree`
//!  - `s` is the Adams filtration
//!  - `min_degree` is the minimum degre of $\Sigma^k M$.
//!  - `num_gens` is the number of generators in this Ext group
//!  - `matrix` is the matrix representing the suspension map from $\Sigma^k M$. This is omitted if
//!    the source or target of the suspension map is trivial, or if they have the same dimension
//!    and the matrix is the identity matrix.
//!
//! The output is best read after sorting with `sort -n -k 1 -k 2 -k 3`.

use std::{path::PathBuf, sync::Arc};

use algebra::module::{Module, SuspensionModule};
use algebra::Algebra;
use chart::{Backend, Orientation, TikzBackend};
use ext::chain_complex::{FiniteChainComplex, FreeChainComplex};
use ext::resolution::UnstableResolution;
use sseq::coordinates::Bidegree;

fn main() -> anyhow::Result<()> {
    let module = Arc::new(ext::utils::query_unstable_module_only()?);
    let save_dir = {
        let base = query::optional("Module save directory", |x| {
            core::result::Result::<PathBuf, std::convert::Infallible>::Ok(PathBuf::from(x))
        });
        move |shift| {
            base.as_ref().cloned().map(|mut x| {
                x.push(format!("suspension{shift}"));
                x
            })
        }
    };

    let max = Bidegree::n_s(
        query::raw("Max n", str::parse),
        query::raw("Max s", str::parse),
    );

    let disp_template: String = query::raw(
        "LaTeX name template (replace % with min degree)",
        str::parse,
    );

    let products = module.algebra().default_filtration_one_products();

    for shift_t in 0..max.n() - module.min_degree() + 3 {
        let shift = Bidegree::s_t(0, shift_t);
        let res: Arc<UnstableResolution<FiniteChainComplex<_>>> =
            Arc::new(UnstableResolution::new_with_save(
                Arc::new(FiniteChainComplex::ccdz(Arc::new(SuspensionModule::new(
                    Arc::clone(&module),
                    shift.t(),
                )))),
                save_dir(shift.t()),
            )?);

        res.compute_through_stem(max + shift);

        println!("\\begin{{figure}}[p]\\centering");

        let sseq = res.to_sseq();
        let products = products
            .iter()
            .map(|(name, op_deg, op_idx)| {
                (name.clone(), res.filtration_one_products(*op_deg, *op_idx))
            })
            .collect::<Vec<_>>();

        sseq.write_to_graph(
            TikzBackend::new(std::io::stdout()),
            2,
            false,
            products.iter(),
            |g| {
                g.text(
                    1,
                    max.s() as i32 - 1,
                    disp_template.replace('%', &format!("{shift_t}")),
                    Orientation::Right,
                )
            },
        )?;

        println!("\\end{{figure}}");
    }
    Ok(())
}

File: ext_py/examples/unstable_chart.py
#!/usr/bin/env python3
"""Compute and chart the suspension maps between unstable Ext groups.

Given an unstable Steenrod module M, compute the unstable Ext groups of the
suspensions of M for all shifts up to the stable range, writing a TikZ figure
for each shift to stdout.

Python port of ext/examples/unstable_chart.rs.
"""

import os
import sys

import _query as query
import ext
from ext import algebra, sseq


def query_unstable_module_only():
    """Inline mirror of ext::utils::query_unstable_module_only.

    Queries a single "Module" spec, parses the optional ``@adem``/``@milnor``
    algebra suffix (default Milnor) and the module name, builds the algebra with
    ``unstable=True`` and returns the corresponding Steenrod module.
    """

    def parse_spec(spec):
        # Mirror Config::try_from(&str): split on '@' for the algebra type.
        module_name, _, algebra_name = spec.partition("@")
        if algebra_name == "":
            algebra_type = algebra.AlgebraType.Milnor
        elif algebra_name == "adem":
            algebra_type = algebra.AlgebraType.Adem
        elif algebra_name == "milnor":
            algebra_type = algebra.AlgebraType.Milnor
        else:
            raise ValueError(f"Invalid algebra type: {algebra_name}")
        # NOTE: depends on ext.parse_module_name (API_PROPOSAL §7.4).
        module = ext.parse_module_name(module_name)
        return (module, algebra_type)

    module_json, algebra_type = query.raw("Module", parse_spec)
    alg = algebra.SteenrodAlgebra.from_json(module_json, algebra_type, True)
    return algebra.SteenrodModule.from_spec(module_json, alg)


def main():
    module = query_unstable_module_only()

    # Mirror the `save_dir` closure: an optional base directory under which each
    # shift gets its own `suspension{shift}` subdirectory.
    base = query.optional("Module save directory", str)

    def save_dir(shift):
        if base is None:
            return None
        return os.path.join(base, f"suspension{shift}")

    max = sseq.Bidegree.n_s(
        query.raw("Max n", int),
        query.raw("Max s", int),
    )

    disp_template = query.raw(
        "LaTeX name template (replace % with min degree)",
        str,
    )

    products = module.algebra().default_filtration_one_products()

    for shift_t in range(0, max.n - module.min_degree() + 3):
        shift = sseq.Bidegree.s_t(0, shift_t)
        # NOTE: depends on ext.SuspensionModule, ext.ChainComplex.ccdz and
        # ext.UnstableResolution.new_with_save (API_PROPOSAL §7.1, §7.3, §7.5).
        res = ext.UnstableResolution(
            ext.ChainComplex.ccdz(
                algebra.SuspensionModule(module, shift.t)
            ),
            save_dir=save_dir(shift.t),
        )

        res.compute_through_stem(max + shift)

        print("\\begin{figure}[p]\\centering")

        ss = res.to_sseq()
        shift_products = [
            (name, res.filtration_one_products(op_deg, op_idx))
            for (name, op_deg, op_idx) in products
        ]

        def header(g, shift_t=shift_t):
            return g.text(
                sseq.Bidegree.x_y(1, max.s - 1),
                disp_template.replace("%", f"{shift_t}"),
                sseq.Orientation.Right,
            )

        ss.write_to_graph(
            sseq.TikzBackend(sys.stdout),
            2,
            False,
            shift_products,
            header,
        )

        print("\\end{figure}")


if __name__ == "__main__":
    main()
